### DISCUSSION TITLE
Application provided MSC class buffer

### DISCUSSION THREAD

**Discussion by raphaelcoeffic** [hathach/tinyusb#2243] [Q&A] [answered]
As of now, it seems that the MSC class buffer is statically allocated directly in `msc_device.c` as follows:
```c
CFG_TUD_MEM_SECTION CFG_TUSB_MEM_ALIGN tu_static uint8_t _mscd_buf[CFG_TUD_MSC_EP_BUFSIZE];
```

The target application for which I'd like to use TinyUSB (https://github.com/EdgeTX/edgetx) currently uses the old ST StdPeriph based USB middleware and we are able with slight modifications to use an application provided buffer.

The use case is as follows:
- when working in USB MSC device mode, the application does not use some memory chunks which is then used / assigned to the USB middleware.
- in that mode, there is no user interaction so that we re-use a memory chunk that is otherwise used by the UI when running in "normal" mode (either no USB connection, or USB HID, or USB CDC).

With the current static MSC class buffer in TinyUSB, that makes a whooping 512 byte sitting idle while we could really use it otherwise.

So I guess the question is: are there any plans to allow for application allocated buffers to be used rather than relying solely on a private statically allocated buffer used solely for that purpose?

**Comment by Rocky04** [comment, marked as answer]
For the same reason I rewrote the HID class driver to use the given buffer directly. Especially when multiple interfaces of a class are used with different transfer sizes it can cause a lot of wasted memory. Also maybe not all interfaces even need a dedicated transfer buffer.

If you don't want to rewrite the MSC class yourself the only solution would be to split up the transfer into multiple transfers by your application. Like to utilize the transfer complete callback and then send the next chunk of data until everything was sent.

A rewrite is especially very useful if the endpoint size (endpoint parameter `wMaxPacketSize`) is big enough to handle the entire transfer in just a single transaction. Then the transfer buffer would be completely wasted because it wouldn't be needed at all. _Be aware that besides its naming the `CFG_TUD_MSC_EP_BUFSIZE` define don't specify the endpoint size and so the transaction size but instead the transfer size._

_Edit:_
So to answer your question, I don't think there are any plans for this.

**Reply by raphaelcoeffic** [reply]
Thanks a lot for the details answer. I will check whether I could spend some time doing this work. Is that something that would have a chance of getting accepted?

**Comment by Rocky04** [comment]
I'm not a TinyUSB maintainer, so I can't tell. I assume that because all class drivers use static allocated transfer buffers.

I haven't created a PR of my custom HID class driver example yet. For now I still handle it as a draft. If you want to check it out, [here is the link](https://github.com/Rocky04/tinyusb/tree/chidd).
